liberum_core/src/swarm_runner/behaviour/kademlia.rs

```diff
@@ -1,87 +1,353 @@
+                qid = format!("{id}"),
+                "Put Record Progressed: Channel closed"
+            );
+        }
+    }
+
+    fn handle_outbound_query_progressed_get_providers_found(
+        &mut self,
+        id: QueryId,
+        _key: RecordKey,
+        providers: HashSet<PeerId>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(providers).inspect_err(|e| {
+                debug!(
+                    node = self.node.name,
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+            // Finish the query to prevent from triggering FinishedWithNoAdditionalRecord
+            self.swarm
+                .behaviour_mut()
+                .kademlia
+                .query_mut(&id)
+                .unwrap()
+                .finish();
+        }
+    }
+
+    fn handle_outbound_query_progressed_get_providers_no_additional_record(
+        &mut self,
+        id: QueryId,
+        _closest_peers: Vec<PeerId>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        debug!(
+            node = self.node.name,
+            "Get providers didn't find any new records"
+        );
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(HashSet::new()).inspect_err(|e| {
+                debug!(
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+        }
+    }
+}
+
+/// Methods to handle Kademlia InboundRequest events
+impl SwarmContext {
+    fn handle_inbound_request_put_record(
+        &mut self,
+        _source: PeerId,
+        _connection: ConnectionId,
+        record: Option<Record>,
+    ) {
+        debug!(node = self.node.name, "Kad Received PutRecord");
+        if record.is_none() {
+            warn!("Received PutRecord with no record");
+            return;
+        }
+        let record = record.unwrap();
+        let id = record.key.clone();
+
+        // save record to filem should use a VAULT here instead
+        self.put_record_into_vault(record);
+
+        // Start a query to be providing the file ID in kademlia
+        let qid = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .start_providing(id.clone());
+        if qid.is_err() {
+            debug!(
+                node = self.node.name,
+                err = format!("{qid:?}"),
+                "Failed to start providing file"
+            );
+            return;
+        }
+        let qid = qid.unwrap();
+
+        // We can't await for a response here because it would block the event loop
+        let (response_sender, _) = oneshot::channel();
```

Comment:
We are dropping receiver here, don't we want to get an answer?

Reply:
We cannot because the answer can only be received from an event handler, which won't be called if the event loop is waiting. A rework of the swarm logic should probably be made to use Arc Mutii and spawn tasks to handle the events instead of blocking the event loop to handle every event.

Reply:
Let's do PR for that.